src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Helpers.cs

```diff
@@ -210,6 +210,35 @@ public static void ValidateDer(ReadOnlyMemory<byte> encodedValue)
         }
     }
 
+    internal static class DictionaryStringHelper
+    {
+        internal static string ReadDirectoryString(this AsnReader tavReader)
+        {
+            Asn1Tag tag = tavReader.PeekTag();
+
+            if (tag.TagClass != TagClass.Universal)
+            {
+                throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+            }
+
+            switch ((UniversalTagNumber)tag.TagValue)
+            {
+                case UniversalTagNumber.BMPString:
+                case UniversalTagNumber.IA5String:
+                case UniversalTagNumber.PrintableString:
+                case UniversalTagNumber.UTF8String:
+                case UniversalTagNumber.T61String:
+                    // .NET's string comparisons start by checking the length, so a trailing
+                    // NULL character which was literally embedded in the DER would cause a
+                    // failure in .NET whereas it wouldn't have with strcmp.
+                    return tavReader.GetCharacterString((UniversalTagNumber)tag.TagValue).TrimEnd('\0');
+                    
+                default:
+                    throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+            }
+        }
+    }
+
```

Comment:
The whole `ReadDirectoryString` method is just moved around from `X500NameEncoder`. The added `.TrimEnd('\0')` is needed to pass unit test, it was previously done inside `DerSequenceReader`. Technically `IA5String` is also not part of DictionaryString, but making that correction is outside the scope of the PR.

Reply:
`ReadDirectoryOrIA5String`?.  `ReadTrimmedDirectoryOrIA5String`?

The TrimEnd certainly matches what I did in DerSequenceReader... though it clearly should have only read up to the first '\0', not just trimmed ones off the end.  I'm sort of on the fence of whether we should leave it as is or fix it to match the comment.  But probably as a followup, since it bothers me to have functional changes and refactorings in the same checkin.  :smile:

Reply:
I'd prefer to make a follow up PR to clean this up. I have some ideas on how to make it cleaner :-)